lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -665,14 +687,147 @@ def write(chunk)
+    #     "details": [
+    #       {
+    #         "@type": "type.googleapis.com/google.logging.v2.WriteLogEntriesPar
+    #           tialErrors",
+    #         "logEntryErrors": {
+    #           "0": {
+    #             "code": 7,
+    #             "message": "User not authorized."
+    #           },
+    #           "1": {
+    #             "code": 3,
+    #             "message": "Log name contains illegal character :"
+    #           },
+    #           "3": {
+    #             "code": 3,
+    #             "message": "Log name contains illegal character :"
+    #           }
+    #         }
+    #       },
+    #       {
+    #         "@type": "type.googleapis.com/google.rpc.DebugInfo",
+    #         "detail": ......
+    #       }
+    #     ]
+    #   }
+    # }
+    #
+    # Note that the root level "code", "message" and "status" simply matches the
+    # root cause of the first failed log entry. For example, if we switch the
+    # order of the log entries, then we will get:
+    # {
+    #    "error" : {
+    #       "code" : 400,
+    #       "message" : "Log name contains illegal character :",
+    #       "status" : "INVALID_ARGUMENT",
+    #       "details": ......
+    #    }
+    # }
+    # We will ignore it anyway and look at the details instead which includes
+    # info for all failed log entries.
+    #
+    # In this example, the logEntryErrors we are extracting should be:
+    # {
+    #   "0": {
+    #     "code": 7,
+    #     "message": "User not authorized."
+    #   },
+    #   "1": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   },
+    #   "3": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   }
+    # }
+    def extract_log_entry_errors(error)
+      error_details = JSON.parse(error.body)['error']['details']
+      partial_errors = error_details.detect do |error_detail|
+        error_detail['@type'] == PARTIAL_ERROR_FIELD
+      end
+      partial_errors['logEntryErrors']
+    rescue JSON::ParserError => e
+      @log.warn 'Failed to extract log entry errors from the error details' \
+                " because of JSON parsing error: #{error.body}.", error: e
+      {}
+    rescue => e
+      @log.warn 'Failed to extract log entry errors from the error details' \
+                " JSON: #{error.body}.", error: e
+      {}
+    end
+
+    # Given the logEntryErrors, constuct a map from errors to a list of indexes
+    # of log entries that failed for this specific error.
+    #
+    # The keys in this map are an array of [error_code, error_message]. The
+    # values in this map are a list of indexes of log entries that failed due to
+    # this error.
+    #
+    # For logEntryErrors as in the example above (in the comment section of the
+    # function extract_log_entry_errors), the constructed map would be:
+    # {
+    #   [
+    #     7,
+    #     'User not authorized.'
+    #   ]: ["0"],
```

Comment:
This is a Ruby hash, not a JSON object, so let's use single quotes around the strings. In fact, the whole thing would fit on a couple of lines:
```ruby
    # {
    #   [7, 'User not authorized.']: ['0'],
    #   [3, 'Log name contains illegal character :']: ['1', '3']
    # }
```

Reply:
Changed.